specs/logging_implementation.adoc

```diff
@@ -0,0 +1,284 @@
+* Logs and alerts need to be treated as messages emitted by Tendrl modules.
+* A schema needs to be defined for emitted messages.
+* Message priorities, types etc. need to be defined at the time of emission.
+* The integration/bridge applications must pass the message as per schema.
+* For messages that serve as notifications to the administrator (or alerts),
+  exact error message must be hardcoded by the developers. No backtraces or
+  system generated error messages should be sent as notifications to the
+  administrator.
+* Notifications sent to the administrator should provide sufficient context to
+  enable debugging of the problem via investigation into the log files.
+* Notifications and alerts need to be delivered to the node agent alert sockets.
+* Common module needs to implement a logging library that hides all the implementation
+  details and configuration options from any of the components using the library.
+  Components should be able to blindly emit messages to the Message Handler module
+  in Tendrl and it should be able to deliver them to either syslog, log files,
+  alert sockets etc. depending upon the metadata and configuration.
+* Syslog must be the default logging configuration. Logging to individual
+  configuration files should be thoroughly discouraged.
+
+== Alternatives
+
+* None
+
+== Data model impact
+
+None
+
+== Impacted modules
+
+=== Tendrl/common impact
+* Create a new class called "Message" in common for message handling with attributes
+  handler_name, message_type, message, log_level, timestamp(format: ISO 8601).
+* Create a new class called "Log" in common and inherit it from Message class.
+* Create a new class called "Alert" in common with attributes alert_id, node_id,
+  resource, current_value, tags, type, severity, significance, ackedby, acked.
+  Inherit alert class from Message class.
+```
+If it is alert then check it is clear alert or not.
+ - Pass the alert details to the function which is declared in socket handler module to
+   check it is clear alert or not. If yes then log the alert as INFO otherwise log the
+   alert based on severity.
+ - loglevel for alert is based on alert severity
+   - If alert severity is Warning then loglevel is WARNING.
+   - If alert severity is Critical then loglevel is CRITICAL.
+   - If alert severity is Ok then loglevel is INFO.
+```
+* Create a new class called "InFlightOperation" in common with attributes
+  job_id, operation_summary. Inherit InFlightOperation class from Message class.
+```
+Alert class object structure
+  - alert_id
+  - node_id
+  - resource
+  - current_value
+  - tags
+  - type
+  - severity
+  - significance
+  - ackedby
+  - acked
+
+Flight operation class object structure
+  - job_id
+  - operation_summary
+
+Log class object structure
+  - It contains no atrribute, it is based on super class(message). For more info
+    see the sample implementation at the end.
+
+Message class object structure
+  - log_handler_name
```

Comment:
@brainfunked we want to log the massage from common, So log message is logged using modules logging configuration. (e.g) if ceph_integration call massage handler to log some message then log message should goes to ceph_integration log file. so these logging handler is passed from each module with message to log the message in corresponding log file.